pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,339 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"strings"
+
+	networking "istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/util/sets"
+)
+
+func mergeVirtualServicesIfNeeded(vServices []Config) (out []Config) {
+	out = make([]Config, 0, len(vServices))
+	delegatesMap := map[string]Config{}
+	// root virtualservices with delegate
+	var rootVses []Config
+
+	// 1. classify virtualservices
+	for _, vs := range vServices {
+		rule := vs.Spec.(*networking.VirtualService)
+		// it is delegate, add it to the indexer cache
+		if len(rule.Hosts) == 0 {
+			delegatesMap[key(vs.Name, vs.Namespace)] = vs
+			continue
+		}
+
+		// root vs
+		if isRootVs(rule) {
+			rootVses = append(rootVses, vs)
+			continue
+		}
+
+		// the others are normal vs without delegate
+		out = append(out, vs)
+	}
+
+	// If `PILOT_ENABLE_VIRTUAL_SERVICE_DELEGATE` feature disabled,
+	// filter out invalid vs(root or delegate), this can happen after enable -> disable
+	if !features.EnableVirtualServiceDelegate.Get() {
+		return
+	}
+
+	// 2. merge delegates and root
+	for _, root := range rootVses {
+		rootVs := root.Spec.(*networking.VirtualService)
+		mergedRoutes := []*networking.HTTPRoute{}
+		for _, route := range rootVs.Http {
+			// it is root vs with delegate
+			if route.Delegate != nil {
+				delegate, ok := delegatesMap[key(route.Delegate.Name, route.Delegate.Namespace)]
+				if !ok {
+					log.Debugf("delegate virtual service %s/%s of %s/%s not found",
+						route.Delegate.Namespace, route.Delegate.Name, root.Namespace, root.Name)
+					// delegate not found, ignore only the current HTTP route
+					continue
+				}
+				vs := delegate.Spec.(*networking.VirtualService)
+				merged := mergeHTTPRoute(route, vs.Http)
+				mergedRoutes = append(mergedRoutes, merged...)
+			} else {
+				mergedRoutes = append(mergedRoutes, route)
+			}
+		}
+		rootVs.Http = mergedRoutes
+		out = append(out, root)
+	}
+
+	return
+}
+
+// merge root's route with delegate's and the merged route number equals the delegate's.
+// if there is a conflict with root, the route is ignored
+func mergeHTTPRoute(root *networking.HTTPRoute, delegate []*networking.HTTPRoute) []*networking.HTTPRoute {
+	root.Delegate = nil
+
+	out := make([]*networking.HTTPRoute, 0, len(delegate))
+	for _, subRoute := range delegate {
+		// suppose there are N1 match conditions in root, N2 match conditions in delegate
+		// if match condition of N2 is a subset of anyone in N1, this is a valid matching conditions
```

Comment:
doesnt that beat the purpose of delegation? The owner of root VS simply says /api/products should be delegated to vs1. vs1 can have any number of rules like /api/products/foo + few header matches. Did you mean all conditions in N1 must be a subset of N2 ?

Reply:
It is that all conditions of the delegate must be a subset of any of root's. Because now all the conditions logic is or, if anyone of the condition matched, it is routable.